File: sdk/python/src/agent_hooks/emitter.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
"""Host-side emitter: dispatch context → interceptors → verdict → enforce (§6, §7, §8, §9)."""
from __future__ import annotations

import copy
import inspect
from typing import Any

from agent_hooks._types import (
    ALLOW,
    Decision,
    EnforcementMode,
    HostError,
    InterceptionPoint,
    InterceptionRecord,
    Verdict,
)
from agent_hooks.approval import (
    ApprovalOutcome,
    ApprovalRequest,
    ApprovalResolver,
)
from agent_hooks.canonical import context_identity
from agent_hooks.context import AgentContext
from agent_hooks.exceptions import InterceptionBlocked
from agent_hooks.interceptor import Interceptor
from agent_hooks.path import PathError
from agent_hooks.path import apply as apply_path


class InterceptionEmitter:
    """Host-side helper that implements §6–§9 once so adapters don't have to.

    An adapter constructs one :class:`InterceptionEmitter` per session, registers
    interceptors and an optional approval resolver, then calls :meth:`emit` at
    each interception point with a built :class:`AgentContext`. The emitter:

    1. Computes ``input_identity`` (§10.2).
    2. Dispatches to interceptors in registration order, combining per §7.1.
    3. On ``escalate``, consults the resolver (§9).
    4. On ``transform`` in ``enforce`` mode, applies the transform to
       ``ctx["target"]`` (§5.2, §6).
    5. Computes ``enforced_identity`` and returns a :class:`InterceptionRecord`.

    The emitter does NOT raise on block verdicts; the adapter inspects
    :attr:`InterceptionRecord.proceeds` (or calls :meth:`emit_or_raise`).
    """

    __slots__ = ("_interceptors", "_mode", "_resolver", "_results")

    def __init__(
        self,
        *,
        mode: EnforcementMode = EnforcementMode.ENFORCE,
        resolver: ApprovalResolver | None = None,
    ) -> None:
        self._interceptors: list[Interceptor] = []
        self._resolver = resolver
        self._mode = mode
        self._results: list[InterceptionRecord] = []

    @property
    def mode(self) -> EnforcementMode:
        return self._mode

    @property
    def results(self) -> list[InterceptionRecord]:
        """All interception records emitted so far in this session, in order."""
        return list(self._results)

    def register(self, interceptor: Interceptor) -> InterceptionEmitter:
        self._interceptors.append(interceptor)
        return self

    # -------------------------------------------------------------------------

    async def emit(self, ctx: AgentContext) -> InterceptionRecord:
        hp = InterceptionPoint(ctx["interception_point"])
        input_id = context_identity(ctx)
        verdict = await self._dispatch(ctx)

        if verdict.decision is Decision.ESCALATE and self._mode is EnforcementMode.ENFORCE:
            verdict = self._resolve_escalate(hp, ctx, verdict, input_id)

        transformed: Any | None = None
        enforced_id = input_id
        if verdict.decision is Decision.TRANSFORM:
            if not hp.transform_permitted:
                verdict = Verdict.host_error(HostError.TRANSFORM_TARGET_FORBIDDEN)
            elif self._mode is EnforcementMode.ENFORCE:
                try:
                    transformed = apply_path(
                        copy.deepcopy(ctx["target"]),
                        verdict.transform.path,  # type: ignore[union-attr]
                        verdict.transform.value,  # type: ignore[union-attr]
                    )
                except PathError as e:
                    verdict = Verdict.host_error(e.host_error, str(e))
                else:
                    ctx["target"] = transformed
                    self._write_back_target(hp, ctx, transformed)
                    enforced_id = context_identity(ctx)
            else:  # evaluate_only — validate only (§8)
                try:
                    apply_path(
                        copy.deepcopy(ctx["target"]),
                        verdict.transform.path,  # type: ignore[union-attr]
                        verdict.transform.value,  # type: ignore[union-attr]
                    )
                except PathError as e:
                    verdict = Verdict.host_error(e.host_error, str(e))

        result = InterceptionRecord(
            interception_point=hp,
            mode=self._mode,
            verdict=verdict,
            input_identity=input_id,
            enforced_identity=enforced_id,
            transformed_target=transformed,
        )
        self._results.append(result)
        return result

    async def emit_or_raise(self, ctx: AgentContext) -> InterceptionRecord:
        """:meth:`emit`, then raise :class:`InterceptionBlocked` if the action must not proceed."""
        result = await self.emit(ctx)
        if not result.proceeds:
            raise InterceptionBlocked(result)
        return result

    # -------------------------------------------------------------------------

    async def _dispatch(self, ctx: AgentContext) -> Verdict:
        """Combine interceptor verdicts per §7.1."""
        combined = ALLOW
        for c in self._interceptors:
            try:
                raw = c.intercept(ctx)
                if inspect.isawaitable(raw):
                    raw = await raw
                v = raw if isinstance(raw, Verdict) else Verdict.from_wire(raw)
            except ValueError as e:
                return Verdict.host_error(HostError.VERDICT_INVALID, str(e))
            except Exception as e:  # noqa: BLE001 — fail closed per §6.3
                return Verdict.host_error(HostError.INTERCEPTOR_FAILED, repr(e))
            if v.decision.blocks:
                return v  # first block short-circuits (§7.1.2)
            if v.decision is Decision.TRANSFORM:
                combined = v  # last transform wins (§7.1.3)
            elif v.decision is Decision.WARN and combined.decision is Decision.ALLOW:
                combined = v
        return combined

    def _resolve_escalate(
        self, hp: InterceptionPoint, ctx: AgentContext, verdict: Verdict, identity: str
    ) -> Verdict:
        if self._resolver is None:
            return Verdict.host_error(HostError.APPROVAL_RESOLVER_MISSING)
        try:
            res = self._resolver.resolve(
                ApprovalRequest(
                    context_identity=identity, interception_point=hp, verdict=verdict, context=ctx
                )
            )
        except Exception as e:  # noqa: BLE001
            return Verdict.host_error(HostError.APPROVAL_RESOLVER_FAILED, repr(e))
        if res.context_identity != identity:
            return Verdict.host_error(HostError.APPROVAL_ACTION_MISMATCH)
        if res.outcome is ApprovalOutcome.UNRESOLVED or res.verdict is None:
            return Verdict.host_error(HostError.APPROVAL_UNRESOLVED)
        return res.verdict

    @staticmethod
    def _write_back_target(hp: InterceptionPoint, ctx: AgentContext, transformed: Any) -> None:
        """Mirror the transformed target back into the L1 field it aliases (§4.3).

        ``target`` is defined as a *reference* to the L1 payload, but Python
        adapters typically build them as separate objects; this keeps both in
        sync so the action consumes the transformed value.
        """
        match hp:
            case InterceptionPoint.INPUT:
                ctx["input"] = transformed
            case InterceptionPoint.PRE_MODEL_CALL:
                ctx["messages"] = transformed
            case InterceptionPoint.POST_MODEL_CALL:
                ctx["response"] = transformed
            case InterceptionPoint.PRE_TOOL_CALL:
                ctx["tool_call"]["args"] = transformed
            case InterceptionPoint.POST_TOOL_CALL:
                ctx["tool_result"]["value"] = transformed
            case InterceptionPoint.OUTPUT:
                ctx["output"] = transformed
            case _:
                pass  # transform_permitted=False already handled

File: sdk/rust/core/src/emitter.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
//! Host-side emitter for Rust-native hosts: dispatch context →
//! interceptors → verdict → record (§6–§9).
//!
//! Unlike the FFI-bound SDKs, this emitter calls the crate's primitives
//! directly (no JSON round-trip). Semantics are identical and pinned by
//! the same CTK vectors:
//!
//! §7.1 sequential fold-through: interceptors run in registration order;
//! each receives its own copy of the context as it stands *after* prior
//! transforms were applied. The first block verdict short-circuits.
//!
//! Fail-closed defaults: an `enforce`-mode emission with zero registered
//! interceptors yields `deny host_error:no_interceptor` (§7), and
//! [`InterceptionEmitter::emit`] returns `Err(InterceptionBlocked)` on
//! any block — the ignorable-record variant is the explicitly named
//! [`InterceptionEmitter::emit_unchecked`].
//!
//! # Timeouts (§7)
//!
//! Unlike the Python/TypeScript/.NET/Go emitters, this emitter does
//! **not** enforce the §7 RECOMMENDED 5000 ms interceptor/resolver
//! timeout: the crate is runtime-agnostic (no tokio/async-std
//! dependency), and a portable future timeout requires a timer driver.
//! Rust hosts own the timeout at the interceptor boundary — wrap each
//! implementation with the host runtime's timeout and map the breach to
//! the reserved reasons, e.g. under tokio:
//!
//! ```ignore
//! // inside your Interceptor::intercept impl
//! match tokio::time::timeout(Duration::from_millis(5000), inner.intercept(ctx)).await {
//!     Ok(v) => v,
//!     Err(_) => Verdict::host_error(HostError::InterceptorTimeout, None),
//! }
//! ```
//!
//! The wrapped verdict flows through the normal §6.3 fail-closed path.

use crate::canonical::context_identity;
use crate::enforce::{apply_transform_to_ctx, finalize, validate_transform};
use crate::types::{
    AgentContext, ApprovalOutcome, ApprovalRequest, ApprovalResolver, Decision, EnforcementMode,
    HostError, Interceptor, InterceptionPoint, InterceptionRecord, Verdict,
};
use serde_json::Value;
use std::fmt;

/// Returned by [`InterceptionEmitter::emit`] when a verdict blocks the
/// guarded action (§6).
#[derive(Debug, Clone)]
pub struct InterceptionBlocked {
    pub record: InterceptionRecord,
}

impl fmt::Display for InterceptionBlocked {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{} blocked: {} ({})",
            self.record.interception_point.as_str(),
            self.record.verdict.decision.as_str(),
            self.record.verdict.reason.as_deref().unwrap_or("no reason"),
        )
    }
}

impl std::error::Error for InterceptionBlocked {}

/// Whether a verdict was synthesized by the host (§11) rather than
/// returned by an interceptor or resolver.
fn is_host_synthesized(v: &Verdict) -> bool {
    v.reason
        .as_deref()
        .is_some_and(|r| r.starts_with("host_error:"))
}

/// Host-side helper that implements §6–§9 once so adapters don't have
/// to. One instance per session.
pub struct InterceptionEmitter {
    interceptors: Vec<Box<dyn Interceptor>>,
    resolver: Option<Box<dyn ApprovalResolver>>,
    mode: EnforcementMode,
    records: Vec<InterceptionRecord>,
}

impl InterceptionEmitter {
    pub fn new(mode: EnforcementMode, resolver: Option<Box<dyn ApprovalResolver>>) -> Self {
        Self {
            interceptors: Vec::new(),
            resolver,
            mode,
            records: Vec::new(),
        }
    }

    pub fn mode(&self) -> EnforcementMode {
        self.mode
    }

    /// All interception records emitted so far in this session, in order.
    pub fn records(&self) -> &[InterceptionRecord] {
        &self.records
    }

    pub fn register(&mut self, interceptor: Box<dyn Interceptor>) -> &mut Self {
        self.interceptors.push(interceptor);
        self
    }

    // -------------------------------------------------------------------------

    /// Run the interception and return `Err(InterceptionBlocked)` if the
    /// guarded action must not proceed (§6). Primary entry point; the
    /// safe path is the default.
    pub async fn emit(
        &mut self,
        ctx: &mut AgentContext,
    ) -> Result<InterceptionRecord, InterceptionBlocked> {
        let record = self.emit_unchecked(ctx).await;
        if record.proceeds() {
            Ok(record)
        } else {
            Err(InterceptionBlocked { record })
        }
    }

    /// Run the interception and return the record without a block error.
    /// The caller MUST inspect [`InterceptionRecord::proceeds`] and halt
    /// the guarded action itself; prefer [`Self::emit`].
    pub async fn emit_unchecked(&mut self, ctx: &mut AgentContext) -> InterceptionRecord {
        // §10.2: input identity binds to the context BEFORE dispatch, so
        // neither interceptor mutation nor fold-through can retroactively
        // alter what the record claims was evaluated.
        let input_id = context_identity(ctx);

        let (mut verdict, mut decided_by) = self.dispatch(ctx).await;

        // §6.1a: nothing to approve at agent_shutdown.
        let at_shutdown = ctx.get("interception_point").and_then(serde_json::Value::as_str)
            == Some("agent_shutdown");
        if verdict.decision == Decision::Escalate
            && self.mode == EnforcementMode::Enforce
            && !at_shutdown
        {
            // §9/NOW-14: approval binds to the escalation-time identity
            // (post prior fold transforms) — what the resolver actually sees.
            let escalation_id = crate::context_identity(ctx);
            verdict = self.resolve_escalate(ctx, verdict, &escalation_id).await;
            // An approve MAY carry a transform (§9); it is subject to the
            // same fold rules as an interceptor transform.
            if verdict.decision == Decision::Transform {
                verdict = self.fold_transform(ctx, verdict);
            }
            // A resolver-substituted verdict keeps the escalating
            // interceptor's index; host-synthesized failures do not.
            if is_host_synthesized(&verdict) {
                decided_by = None;
            }
        }

        let record = finalize(ctx, verdict, self.mode, &input_id, decided_by);
        self.records.push(record.clone());
        record
    }

    // -------------------------------------------------------------------------

    /// §7 dispatch with §7.1 sequential fold-through. Returns the
    /// combined verdict and the registration index of the deciding
    /// interceptor (`None` for pure allow or host-synthesized).
    async fn dispatch(&self, ctx: &mut AgentContext) -> (Verdict, Option<u32>) {
        if self.interceptors.is_empty() {
            // §7: zero interceptors fails closed. Register an explicit
            // allow-all interceptor for a deliberate passthrough.
            return (Verdict::host_error(HostError::NoInterceptor, None), None);
        }

        let mut combined = Verdict::allow();
        let mut decided_by: Option<u32> = None;
        for (i, interceptor) in self.interceptors.iter().enumerate() {
            let i = i as u32;
            // §7.1/N05: each interceptor gets its own copy — in-place
            // mutation of the copy cannot alter enforcement.
            let copy = ctx.clone();
            let v = interceptor.intercept(&copy).await;
            if v.validate().is_err() {
                // §5 gate; the interceptor trait is infallible so the
                // only failure mode is a malformed verdict shape.
                return (Verdict::host_error(HostError::VerdictInvalid, None), None);
            }

            if !v.decision.permits() {
                return (v, Some(i)); // first block short-circuits (§7.1)
            }
            if v.decision == Decision::Transform {
                let v = self.fold_transform(ctx, v);
                if !v.decision.permits() {
                    return (v, None); // transform failed closed (host-synthesized)
                }
                combined = v;
                decided_by = Some(i);
            } else if v.decision == Decision::Warn && combined.decision == Decision::Allow {
                combined = v;
                decided_by = Some(i);
            }
        }
        (combined, decided_by)
    }

    /// Apply (enforce) or validate (evaluate_only) one transform (§7.1, §8).
    fn fold_transform(&self, ctx: &mut AgentContext, v: Verdict) -> Verdict {
        let t = match &v.transform {
            Some(t) => t.clone(),
            None => return Verdict::host_error(HostError::TransformInvalid, None),
        };
        let result = match self.mode {
            EnforcementMode::Enforce => apply_transform_to_ctx(ctx, &t),
            EnforcementMode::EvaluateOnly => validate_transform(ctx, &t),
        };
        match result {
            Ok(()) => v,
            Err(e) => Verdict::host_error(e, Some(t.path)),
        }
    }

    async fn resolve_escalate(
        &self,
        ctx: &AgentContext,
        verdict: Verdict,
        identity: &str,
    ) -> Verdict {
        let Some(resolver) = &self.resolver else {
            return Verdict::host_error(HostError::ApprovalResolverMissing, None);
        };
        let ip: InterceptionPoint = ctx
            .get("interception_point")
            .and_then(Value::as_str)
            .and_then(|s| s.parse().ok())
            .unwrap_or(InterceptionPoint::AgentStartup);
        let res = resolver
            .resolve(ApprovalRequest {
                context_identity: identity.to_owned(),
                interception_point: ip,
                verdict: &verdict,
                context: ctx,
            })
            .await;
        if res.context_identity != identity {
            return Verdict::host_error(HostError::ApprovalActionMismatch, None);
        }
        let Some(rv) = res.verdict else {
            return Verdict::host_error(HostError::ApprovalUnresolved, None);
        };
        if res.outcome == ApprovalOutcome::Unresolved {
            return Verdict::host_error(HostError::ApprovalUnresolved, None);
        }
        // §9/N04: the resolver's verdict crosses the same §5 gate as an
        // interceptor's.
        if rv.validate().is_err() {
            return Verdict::host_error(HostError::VerdictInvalid, None);
        }
        rv
    }
}
